File: patterns/java/src/refactoring_guru/patterns/proxy/example/media_framework_lib/ThirdPartyYoutubeClass.java
package refactoring_guru.patterns.proxy.example.media_framework_lib;

import java.util.ArrayList;
import java.util.List;

public class ThirdPartyYoutubeClass implements ThirdPartyYoutubeLib {
    private List<String> list = new ArrayList<>();

    @Override
    public List<String> listVideos() {
        System.out.println("Connecting to http://www.youtube.com");
        delay();
        System.out.println("\n" + "Connected" + "\n");
        System.out.println("Creating video list");
        delay();
        System.out.println("\n" + "Video list was created" + "\n");
        String[] array = new String[] {"Forsage4.avi", "Dog play with ball.mp4",
                "Dancing video.mpq", "Barselona vs RealM.mov",
                "Programing lesson#1.avi", "Tayson vs Levis.mpq" };
        for (String s : array) {
            list.add(s);
        }
        return list;
    }

    public void delay() {
        for (int i = 0; i < 37; i++) {
            try {
                Thread.sleep(100);
                System.out.print("*");
            } catch (InterruptedException ex) {
                ex.printStackTrace();
            }
        }
    }

    @Override
    public String getVideoInfo(int id) {
        String result = list.get(id - 1);
        String info = "File name: " +result.substring(0, result.indexOf(".")) + "\n"
                + "Video format: " + result.substring(result.indexOf(".") + 1, result.length());
        return info;
    }

    @Override
    public void downloadVideo(int id) {
        System.out.println("File '" + list.get(id - 1) + "' - was downloaded");
    }
}

File: patterns/cs/src/RefactoringGuru/Proxy/Example/SomeCoolMediaLibrary/ThirdPartyYoutubeClass.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace RefactoringGuru.Proxy.Example.SomeCoolMediaLibrary
{
    class ThirdPartyYoutubeClass : ThirdPartyYoutubeLib
    {
        public Dictionary<string, Video> PopularVideos()
        {
            ConnectToServer("http://www.youtube.com");
            return GetRandomVideos();
        }

        public Video GetVideo(string videoId)
        {
            ConnectToServer("http://www.youtube.com/" + videoId);
            return GetSomeVideo(videoId);
        }

        // -----------------------------------------------------------------------
        // Fake methods to simulate network activity. They as slow as a
        // real life.

        private int Random(int min, int max)
        {
            Random rnd = new Random();
            return rnd.Next(min, max);
        }

        private void ExperienceNetworkLatency()
        {
            int randomLatency = Random(5, 10);
            for (int i = 0; i < randomLatency; i++)
            {
                Thread.Sleep(100);
            }
        }

        private void ConnectToServer(String server)
        {
            Console.WriteLine("Connecting to " + server + "... ");
            ExperienceNetworkLatency();
            Console.WriteLine("Connected!" + "\n");
        }

        private Dictionary<String, Video> GetRandomVideos()
        {
            Console.WriteLine("Downloading populars... ");

            ExperienceNetworkLatency();
            Dictionary<String, Video> dictionary = new Dictionary<string, Video>();
            dictionary.Add("catzzzzzzzzz", new Video("sadgahasgdas", "Catzzzz.avi"));
            dictionary.Add("mkafksangasj", new Video("mkafksangasj", "Dog play with ball.mp4"));
            dictionary.Add("dancescideoo", new Video("asdfas3ffasd", "Dancing video.mpq"));
            dictionary.Add("dlsdk5jfslaf", new Video("dlsdk5jfslaf", "Barcelona vs RealM.mov"));
            dictionary.Add("3sdfgsd1j333", new Video("3sdfgsd1j333", "Programing lesson#1.avi"));

            Console.WriteLine("Done!" + "\n");
            return dictionary;
        }

        private Video GetSomeVideo(String videoId)
        {
            Console.WriteLine("Downloading video... ");

            ExperienceNetworkLatency();
            Video video = new Video(videoId, "Some video title");

            Console.WriteLine("Done!" + "\n");
            return video;
        }
    }
}
